pkg/envoy/resource_monitor_linux.go

```diff
@@ -0,0 +1,767 @@
+			&envoy_config_overload_v3.OverloadAction{
+				Name:        fmt.Sprintf("envoy.overload_actions.%s", action.ActionName),
+				Triggers:    []*envoy_config_overload_v3.Trigger{action.Trigger},
+				TypedConfig: overloadActionConfigs[action.ActionName],
+			},
+		)
+	}
+
+	bootstrap.OverloadManager.BufferFactoryConfig = &envoy_config_overload_v3.BufferFactoryConfig{
+		// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/overload/v3/overload.proto#config-overload-v3-bufferfactoryconfig
+		MinimumAccountToTrackPowerOfTwo: 20,
+	}
+}
+
+var (
+	monitorInitialTickDelay = 1 * time.Second
+	monitorMaxTickInterval  = 10 * time.Second
+	monitorMinTickInterval  = 250 * time.Millisecond
+)
+
+func (s *sharedResourceMonitor) Run(ctx context.Context, envoyPid int) error {
+	envoyCgroup, err := s.driver.CgroupForPid(envoyPid)
+	if err != nil {
+		return fmt.Errorf("failed to look up cgroup for envoy process: %w", err)
+	}
+	if envoyCgroup != s.cgroup {
+		return fmt.Errorf("envoy process is not in the expected cgroup: %s", envoyCgroup)
+	}
+	log.Info(ctx).Str("service", "envoy").Str("cgroup", s.cgroup).Msg("starting resource monitor")
+
+	limitWatcher := &memoryLimitWatcher{
+		limitFilePath: "/" + s.driver.Path(s.cgroup, MemoryLimitPath),
+	}
+	lwCtx, lwCancel := context.WithCancel(ctx)
+	defer func() {
+		lwCancel()
+		limitWatcher.Wait()
+	}()
+	if err := limitWatcher.Watch(lwCtx); err != nil {
+		return fmt.Errorf("failed to start watch on cgroup memory limit: %w", err)
+	}
+
+	// Set initial values for state metrics
+	s.updateActionStates(ctx, 0)
+
+	// The interval at which we check memory usage is scaled based on the current
+	// memory saturation. When memory usage is low, we check less frequently, and
+	// as the saturation increases, we also increase the frequency of checks. Most
+	// of the thresholds at which some action is taken to reduce memory usage are
+	// very high (e.g. 95% of the limit). As memory usage approaches this limit,
+	// it becomes increasingly important to have accurate data, since memory usage
+	// can change rapidly; we want to avoid hitting the limit, but also delay
+	// taking disruptive actions for as long as possible.
+
+	// the envoy default interval for the builtin heap monitor is 1s
+
+	tick := time.NewTimer(monitorInitialTickDelay)
+	var lastValue string
+	for {
+		select {
+		case <-ctx.Done():
+			tick.Stop()
+			return ctx.Err()
+		case <-tick.C:
+			var saturation float64
+			if s.enabled.Load() {
+				if limit := limitWatcher.Value(); limit > 0 {
+					usage, err := s.driver.MemoryUsage(s.cgroup)
+					if err != nil {
+						log.Error(ctx).Err(err).Msg("failed to get memory saturation")
+						continue
+					}
+					saturation = float64(usage) / float64(limit)
+				}
+			}
+
+			saturationStr := fmt.Sprintf("%.6f", saturation)
+			nextInterval := (monitorMaxTickInterval - (time.Duration(float64(monitorMaxTickInterval-monitorMinTickInterval) * saturation))).
```

Comment:
maybe we could pull this into a separate function?

Also do we need to handle `saturation > 1`?

Reply:
Yeah I think we should handle that, good catch